Gems/Atom/Asset/Shader/Code/Source/Editor/HashedVariantListSourceData.h

```diff
@@ -0,0 +1,71 @@
+
+namespace AZ
+{
+    class ReflectContext;
+
+    namespace ShaderBuilder
+    {
+        //! This is a simple data structure that represents a .hashedvariantlist file.
+        //! Users are not expected to create .hashedvariantlist files. These files are produced by the ShaderVariantListBuilder
+        //! as intermediate assets.
+        //! Provides configuration data about which shader variants should be generated for a given shader.
+        struct HashedVariantListSourceData
+        {
+            AZ_TYPE_INFO_WITH_NAME_DECL(HashedVariantListSourceData);
+            AZ_CLASS_ALLOCATOR(HashedVariantListSourceData, AZ::SystemAllocator);
+
+            static constexpr const char* Extension = "hashedvariantlist";
+            static constexpr uint32_t SubId = 0;
+
+            static void Reflect(ReflectContext* context);
+
+            //! A struct that describes each hashed shader variant data.
+            struct HashedVariantInfo
+            {
+                AZ_TYPE_INFO_WITH_NAME_DECL(HashedVariantInfo);
+
+                AZ::RPI::ShaderVariantListSourceData::VariantInfo m_variantInfo;
+                size_t m_hash = 0; // Hash of all the data in @m_variantInfo
+                bool m_isNew = false; // If true, the corresponding ShaderVariantAsset should be built/rebuilt.
+
+                //! Hash combines all the data in @optionValues.
+                static size_t HashCombineShaderOptionValues(size_t startingHash, const AZ::RPI::ShaderOptionValuesSourceData& optionValues);
+
+                //! Hash combines the result of a previous call to HashCombineShaderOptionValues, which is passed to this function
+                //! in @optionValuesHash argument, with the rest of the data in @variantInfo.
+                static size_t CalculateHash(size_t optionValuesHash, const AZ::RPI::ShaderVariantListSourceData::VariantInfo& variantInfo);
+
+                //! Same as above, but uses this.m_variantInfo and stores the result in this.m_hash.
+                void CalculateHash(size_t optionValuesHash);
+            };
+
+            // A time stamp is necessary, because building shader variants takes time.
+            // We calculate hashes to figure out the variants that changed and based on the match
+            // we set HashedVariantInfo::m_isNew to false or true.
+            // Imagine a user makes changes to a .shadervariantlist, and some HashedVariantInfos are marked as new.
+            // then within a few seconds later they make another change to the .shadervariantlist file. Because
+            // it happened so quickly those HashedVariantInfos that were set as new would become "old" and those
+            // shader variants won't be compiled.
+            // This timestamp comes to the rescue and we can measure if the change happened too quick
+            // and it such case we preserve the state of each HashedVariantInfo::m_isNew.
+            long long m_timeStamp;
```

Comment:
I'm a little concerned this timestamp concept will have weird edge cases. I wonder if something like this would work, instead of chaining ShaderVariantListBuilder with ShaderVariantAssetBuilder....
Just have a single ShaderVariantListBuilder stand on its own. The first time a list is processed, it will output HashVariantListSourceData to the cache and will also compile all the shader variants. On subsequent updates, it will load the HashVariantListSourceData that it output last time, and check the hashes for changes. It will then compile shaders for only the new/updated variants. You won't need any IsNew flag because the same builder that caches the hashes will also check the hashes and compile the shaders.

Reply:
At face value your idea sounds good. But need to make sure there are no quirks. thinking...

Reply:
See my final conclusion below, as I answer a question from @gadams3